tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/OptimizedSlidingTimeWindowMovingAverages.java

```diff
@@ -0,0 +1,230 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.tritium.metrics.registry;
+
+import com.codahale.metrics.Clock;
+import com.codahale.metrics.ExponentialMovingAverages;
+import com.codahale.metrics.Histogram;
+import com.codahale.metrics.Meter;
+import com.codahale.metrics.MovingAverages;
+import com.codahale.metrics.SlidingTimeWindowArrayReservoir;
+import com.google.common.annotations.VisibleForTesting;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.LongAdder;
+
+/**
+ * Optimized fork of {@link com.codahale.metrics.SlidingTimeWindowMovingAverages}, see
+ * <a href="https://github.com/dropwizard/metrics/pull/4936">upstream PR 4936</a>.
+ * <p>
+ * A triple of simple moving average rates (one, five and fifteen minutes rates) as needed by {@link Meter}.
+ * <p>
+ * The averages are unweighted, i.e. they include strictly only the events in the
+ * sliding time window, every event having the same weight. Unlike the
+ * the more widely used {@link ExponentialMovingAverages} implementation,
+ * with this class the moving average rate drops immediately to zero if the last
+ * marked event is older than the time window.
+ * <p>
+ * A {@link Meter} with {@link com.codahale.metrics.SlidingTimeWindowMovingAverages} works similarly to
+ * a {@link Histogram} with an {@link SlidingTimeWindowArrayReservoir}, but as a Meter
+ * needs to keep track only of the count of events (not the events itself), the memory
+ * overhead is much smaller. SlidingTimeWindowMovingAverages uses buckets with just one
+ * counter to accumulate the number of events (one bucket per seconds, giving 900 buckets
+ * for the 15 minutes time window).
+ */
+public final class OptimizedSlidingTimeWindowMovingAverages implements MovingAverages {
+
+    private static final long TIME_WINDOW_DURATION_MINUTES = 15;
+    private static final long TICK_INTERVAL = TimeUnit.SECONDS.toNanos(1);
+    private static final Duration TIME_WINDOW_DURATION = Duration.ofMinutes(TIME_WINDOW_DURATION_MINUTES);
+
+    // package private for the benefit of the unit test
+    static final int NUMBER_OF_BUCKETS = (int) (TIME_WINDOW_DURATION.toNanos() / TICK_INTERVAL);
+
+    private final AtomicLong lastTick;
+    private final Clock clock;
+
+    /**
+     * One counter per time bucket/slot (i.e. per second, see TICK_INTERVAL) for the entire
+     * time window (i.e. 15 minutes, see TIME_WINDOW_DURATION_MINUTES).
+     */
+    private final LongAdder[] buckets;
```

Comment:
attribution also forked this in https://github.palantir.build/foundry/attribution/blame/develop/attribution-core/src/main/java/com/palantir/attribution/core/SlidingTimeWindowMovingAverages.java